### PR TITLE
Only run checks for intrinsics if optimizations are enabled.

### PR DESCRIPTION
**Pull request by nmoinvaz**
Checks for instrinics supports and intrinsics flags are still being done if `-DWITH_OPTIM=OFF`. I have also cleaned it up and moved the intrinsics checks to `detect-instrinsics.cmake`.

### CODE REVIEW

**Comment by nmoinvaz** [comment]
Still have to do `configure` script..

**Comment by Dead2** [comment]
I like the cmake changes a lot.

The configure changes I am a little unsure about, I feel that they make the file less readable.
Perhaps do the similar thing and make functions?
Should make sure that they are the portable kind though: [with or without function prefix](https://unix.stackexchange.com/questions/73750/difference-between-function-foo-and-foo)

**Comment by nmoinvaz** [comment]
@Dead2 done. The `check_*_intrinsics` functions are sorted alphabetically like in CMake. Also have added a commit that will print out which static/shared object files are used (like CMake does) for verification.

**Comment by Dead2** [comment]
@nmoinvaz Wow, nice! Will do a full review at a later point, but looks good.

File: configure [hunk 1]
```diff
@@ -1466,6 +1470,8 @@ esac
 
 echo "ARCH: ${ARCH}"
 echo "Using arch directory: ${ARCHDIR}"
+echo "Architecture-specific static object files:${ARCH_STATIC_OBJS}"
+echo "Architecture-specific shared object files:${ARCH_SHARED_OBJS}"
```
Comment (mtl1979) [inline configure:1562]: For readability, there should be space after `:`.

Comment (nmoinvaz) [reply configure:1562]: Those variables already have a space at the beginning based on the way they are generated.

File: configure [hunk 2]
```diff
@@ -1262,6 +1259,10 @@ case "${ARCH}" in
         CFLAGS="${CFLAGS} ${floatabi}"
         SFLAGS="${SFLAGS} ${floatabi}"
 
+        if test $without_optimizations -eq 0; then
+            check_neon_intrinsics
+        fi
```
Comment (mtl1979) [inline configure:1337]: We should only check if NEON intrinsics are available if the processor might support NEON or VFP. There is a lot of ARM processors that support neither. We currently support VFPv3 and VFPv4 when compiling with gcc or Visual C++.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1261,7 +1125,6 @@ if(ZLIB_ENABLE_TESTS)
 
     add_simple_test_executable(deflate_quick_bi_valid)
     add_simple_test_executable(deflate_quick_block_open)
-    add_simple_test_executable(hash_head_0)
```
Comment (mtl1979) [inline CMakeLists.txt]: This edit seems out-of-place... Has nothing to do with optimizations.

**Comment by mtl1979** [comment]
Needs rebase.

File: configure (hunk 1 above)
Comment (mtl1979) [reply configure:1562]: At least on modern x86, the lines are really long, over 200 characters. Even with wide-screen aspect ratio, that is very long.

File: CMakeLists.txt (hunk 3 above)
Comment (nmoinvaz) [reply CMakeLists.txt]: Fixed.

File: configure (hunk 2 above)
Comment (nmoinvaz) [reply configure:1337]: I think that would have to be an improvement for another PR. I don't even own a ARM machine.

File: CMakeLists.txt (hunk 3 above)
Comment (nmoinvaz) [reply CMakeLists.txt]: Fixed.

**Comment by nmoinvaz** [comment]
Rebased.

File: configure [hunk 4]
```diff
@@ -1210,13 +1179,17 @@ case "${ARCH}" in
             ARCH_STATIC_OBJS="${ARCH_STATIC_OBJS} x86.o"
             ARCH_SHARED_OBJS="${ARCH_SHARED_OBJS} x86.lo"
 
+            check_avx2_intrinsics
```
Comment (Dead2) [inline configure:1182]: Maybe these could have () at the end? Makes it clear it is a function being called.

**Comment by Dead2** [comment]
Needs rebase

File: configure (hunk 4 above)
Comment (nmoinvaz) [reply configure:1182]: Fixed.

**Comment by nmoinvaz** [comment]
Rebased and added those `()` to configure script.

**Comment by Dead2** [comment]
CI fails with:
```
../configure: line 1337: syntax error near unexpected token `fi'
../configure: line 1337: `        fi'
```

Seems I was wrong in suggesting (), as I find no references to documentation using those. I have always used them myself, but I guess that is actually not portable or even fully functional after all. Learned something there. Sorry about that.

**Comment by nmoinvaz** [comment]
Ok I have switched it back.

**Comment by nmoinvaz** [comment]
Nice!!

**Pull request merged**